### PR TITLE
Add solr client.

### PR DESCRIPTION
**Pull request by pritamdas99**
Signed-off-by: pritamdas99 <pritam@appscode.com>

### CODE REVIEW

File: solr/kubedb_client_builder.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+			return nil, err
+		}
+	}
+	version, err := semver.NewVersion(o.db.Spec.Version)
+	if err != nil {
+		return nil, gerr.Wrap(err, "failed to parse version")
+	}
+
+	switch {
+	case version.Major() == 9:
```
Comment (raihankhan) [inline solr/kubedb_client_builder.go:103]: should be - `version.Major() >= 9:` for future compatability

File: solr/kubedb_client_builder.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	if o.url == "" {
+		o.url = o.GetHostPath(o.db)
+	}
+	if o.db == nil {
+		return nil, errors.New("db is empty")
+	}
+	config := Config{
+		host: o.url,
+		transport: &http.Transport{
+			IdleConnTimeout: time.Minute * 6,
```
Comment (raihankhan) [inline solr/kubedb_client_builder.go:73]: Why do we have 6 minutes (such long periods) of timeouts everywhere? and can we use a variable for it?

File: solr/client.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	"io"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	api "kubedb.dev/apimachinery/apis/kubedb/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Client struct {
+	SLClient
```
Comment (raihankhan) [inline solr/client.go:14]: Can we make this pointer type?

File: solr/kubedb_client_builder.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+		newClient := resty.New()
+		newClient.SetScheme(config.connectionScheme).SetBaseURL(config.host).SetTransport(config.transport)
+		newClient.SetTimeout(6 * time.Minute)
+		newClient.SetHeader("Accept", "application/json")
+		newClient.SetDisableWarn(true)
+		newClient.SetBasicAuth(string(authSecret.Data[core.BasicAuthUsernameKey]), string(authSecret.Data[core.BasicAuthPasswordKey]))
+		return &Client{
+			&SLClientV8{
+				Client: newClient,
+				log:    config.log,
```
Comment (raihankhan) [inline solr/kubedb_client_builder.go:127]: If Config have log inside it, do we need to set log as a separate parameter here? Can't we extract log from config later.

File: solr/kubedb_client_builder.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+			config.log.Error(err, "failed to get auth secret to get solr client")
+			return nil, err
+		}
+	}
+	version, err := semver.NewVersion(o.db.Spec.Version)
+	if err != nil {
+		return nil, gerr.Wrap(err, "failed to parse version")
+	}
+
+	switch {
```
Comment (raihankhan) [inline solr/kubedb_client_builder.go:102]: Do we really need different switch cases for v8 and v9? are there differences in their clients/method implementations ?

File: solr/solrv8.go [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+	writeResponse := &Response{
+		Code:   res.StatusCode(),
+		header: res.Header(),
+		body:   res.RawBody(),
+	}
+	return writeResponse, nil
+}
+
+func (sc *SLClientV8) ReadCollection() (*Response, error) {
+	sc.Config.log.V(5).Info("READING COLLECTION: kubedb-system")
```
Comment (raihankhan) [inline solr/solrv8.go:112]: `kubedb-system` should be used as a constant everywhere

File: solr/util.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
+package solr
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
```
Comment (raihankhan) [inline solr/util.go:7]: Check if the formatting is ok in this import block

**Review by raihankhan — CHANGES_REQUESTED** [review]

File: solr/client.go (hunk 3 above)
Comment (pritamdas99) [reply solr/client.go:14]: can't use pointer for embedded types

File: solr/kubedb_client_builder.go (hunk 5 above)
Comment (pritamdas99) [reply solr/kubedb_client_builder.go:102]: some v2 apis are not supported by version 8

**Review by raihankhan — APPROVED** [review]
LGTM!

**Pull request merged**